# Predicting Annual Salary for Data Science and STEM Roles based on LEVELS.FYI Data

## 1. Business Understanding

What we need to do:
- Determine business objectives
  - Background
  - Business objectives
  - Business success criteria
- Situation Assessment
  - Inventory of resources
  - Requirements, assumptions, and constraints
  - Risks and contingencies
  - Terminology
  - Cost and benefits
- Determine data mining goal
  - Data mining goals
  - Data mining success criteria
- Product project plan
  - Project plan
  - Initial assessment of tools and techniques

### Determine business objectives
<b> Background, overview and motivation: </b>
Tech-related jobs has been booming when technology went mainstream, especially data scientist role.

According to the Glassdoor 2022, data scientist rank the top 3 best job in America for 7 years running. This ranking is based on the career opportunites rating, number of opening jobs, and the average salaries. The median base salary is $120,000. The job opening were also spiking up from 1,736 in 2016 to 10,071 in 2022. 

However, the ecosystem does not depend on just data scientist alone, it involves upstream and downstream job roles - like product/project managers and software engineers. These jobs are equally important to ensure the democratization of technology to average users. Hence, many people flocked to tech industry.

Unfortunately, the compensation structure vary from one company to another, and was not accessible equally to everyone. [Levels.fyi](https://www.levels.fyi/) was started to democratize the access of information to compensation structure across tech jobs.

*We want to expand this effort by creating an AI model (and subsequently a web app) that enable users to know the estimated annual compensation based on user inputs like year, company, level, title, location, years of experience, years at company and gender.*

<b> Business objective: </b>
To predict the annual salary (including base pay, stock grants and estimated bonuses) for data science and STEM roles using data from [LEVELS.FYI](https://www.levels.fyi/).

<b> Business success criteria: </b> 
A web app that enables user to know the annual compensation based on their input like year, company, level, title, location, years of experience, years at company and gender. and for the next 5 years

<b> Related works: </b> 

<b> 1. Salary Prediction in the IT Job Market with Few High-Dimensional Samples: A Spanish Case Study </b><br>
This paper study predict the salary of IT jobs using data from Tecnoempleo, an e-Recruitment website. The result shows that experience, job stability or certain roles contribute significantly to the salary. Besides, decision trees yield a bettery accuracy (84%) than voting committe based tools.

<b> 2. Gender-based salary differences in academic medicine: a retrospective review of data from six public medical centers in the Western USA </b> <br>
The paper study the effects of gender, rank and speciality of the acedemic medical centers in the Western USA. among the 799 faculty members, it is found that male are paid significantly higher than feamle in general surgery and obstetrics and gynaecology. It is estimated female are paid ~$75,000 lower than their male colleague.

### Situation Assessment

xx

### Determine data mining goal and success criteria

Goals:
- **Create a ML model that is able to predict the annual salary based on user input with MAE <= 1000, RMSE <= 1, Pearson >= 0.9, R2 >= 0.9 (Primary outcome)**
- Create a ML model that is able to predict the annual salary for the same role in the next 5 years (Optional)
- Deploy the ML model into a web app which outputs estimated annual salary to the user, which automatically retrains with new data (Optional)

### Project plan

Initial assessment of tools and techniques:
1. Tools: 
   - Pandas
   - Matplotlib
   - Pandas-profiling
   - Scikit-learn
   - Missingno
   - PyTorch (optional)
   - Streamlit (optional)
   - Heroku (optional)
2. Techniques:
   - CRISP-DM
   - Machine learning
   - Deep learning (optional)

### Some interesting questions to be answered during analysis

1. Many people thinks that tech companies pays well above average and cost of living. Is this true?
2. Many people thinks that you don’t need a degree to work in these tech companies. Is this true?
3. Are women underpaid in these companies?
4. Are non-whites paid lower than whites in these companies?
5. Does years in company affect value of granted stock and bonus?
6. For the same position or job level, does location affect base salary regardless of gender and race?
7. What are the factors impacting the salary of a data scientist?
8. What is the salary trend for data scientist? and its comparison to living cost?
9. What is the predicted salary for Data Scientist in next 5 years?

Surprisingly, the data from [Kaggle](https://www.kaggle.com/datasets/jackogozaly/data-science-and-stem-salaries) is different from the latest data pulled from LEVELS.FYI official [JSON API](https://www.levels.fyi/js/salaryData.json) hence we could not answer Question 2, 4 and partially 6 - perhaps to avoid educational and race discrimination.

## 2. Data Understanding

What we need to do:
- Collect initial data
- Describe data
- Explore data
- Verify data quality

### Collect initial data

There are two ways to get the dataset:
1. XLSX file from [Kaggle](https://www.kaggle.com/datasets/jackogozaly/data-science-and-stem-salaries)
2. JSON file from [LEVELS.FYI API](https://www.levels.fyi/js/salaryData.json)

In order to get the latest data, we will use the 2nd method as we want to extend this model into a realtime web app later.

In [20]:
import pandas as pd
import requests
data = requests.get('https://www.levels.fyi/js/salaryData.json').json()
df = pd.DataFrame(data)
df

timestamp               company                level  \
0       6/7/2017 11:33:27                Oracle                   L3   
1      6/10/2017 17:11:29                  eBay                 SE 2   
2      6/11/2017 14:53:57                Amazon                   L7   
3      6/14/2017 21:22:25             Microsoft                   64   
4      6/16/2017 10:44:01                Amazon                   L5   
...                   ...                   ...                  ...   
62637   8/17/2021 8:16:36                Amazon                   L6   
62638   8/17/2021 8:22:17  Fidelity Investments                   L3   
62639   8/17/2021 8:24:56                 Cisco              Grade 8   
62640   8/17/2021 8:26:21                  HSBC                 GCB5   
62641   8/17/2021 8:28:57                 Adobe  Software Engineer 5   

                              title totalyearlycompensation  \
0                   Product Manager                     186   
1                 Software Engineer                       0   
2                   Product Manager                     126   
3      Software Engineering Manager                  171348   
4                 Software Engineer                  214175   
...                             ...                     ...   
62637               Product Manager                     272   
62638             Software Engineer                      50   
62639             Software Engineer                     200   
62640             Software Engineer                      86   
62641              Product Designer                     382   

                    location yearsofexperience yearsatcompany  \
0           Redwood City, CA                 1              1   
1          San Francisco, CA                 3              4   
2                Seattle, WA                10              0   
3                Redmond, WA                13             13   
4      Vancouver, BC, Canada                13              1   
...                      ...               ...            ...   
62637            Seattle, WA                 7              0   
62638             Durham, NC                 0              0   
62639           San Jose, CA                 3              6   
62640           New York, NY                10              5   
62641      San Francisco, CA                 7              0   

                        tag basesalary stockgrantvalue  bonus  gender  \
0                                  160              17      9           
1                                    0               0      0           
2                                  126               0      0           
3                                92799           55371  23178           
4                               156335               0  57840           
...                     ...        ...             ...    ...     ...   
62637              Analytic        176              51     45  Female   
62638            Full Stack         43               0      7    Male   
62639            Networking        179               7     14    Male   
62640            Full Stack         72               0     14           
62641  User Experience (UX)        230             139     13    Male   

                                            otherdetails cityid dmaid  \
0                                                          7392   807   
1                                                          7419   807   
2                                                         11527   819   
3                                                         11521   819   
4                                                          1320     0   
...                                                  ...    ...   ...   
62637  scrielt schwists fuech xauem typaiarty siids q...  11527   819   
62638  pauss jiusts liaop rhiitts proing smiodly splu...   9606   560   
62639                  triasm froidy peusk juieck skaorr   7422   807   
62640 

The dataset from Kaggle is slightly different from the LEVELS.FYI API:
- "Education" has been removed
- "Race" has been removed

Perhaps to avoid educational and race discrimination. Hence, we could not answer Question 2 and 4; while we could only try to answer Question 6 partially.

From the dataframe above, we have the following features:
1. timestamp: Date and time when the details was submitted
2. company: Hiring company
3. level: Job grade
4. title: Job title
5. totalyearlycompensation: Annual salary including base salary, value of granted stock and bonus
6. location: Location of the job
7. yearsofexperience: Total number years of experience
8. yearsatcompany: Total number of years in the company
9. tag: NA
10. basesalary: Basic salary
11. stockgrantvalue: Value of granted stock
12. bonus: Total bonus
13. gender: Gender/sex of the hiree
14. otherdetails: Other details submitted
15. cityid: ID of the city where the job was located
16. dmaid: DMA (Designated Market Area) regions are the geographic areas in the U.S. in which local television viewing is measured by Nielsen
17. rowNumber: Instance ID

### Describe data

Let's first describe the data in simple terms, to establish a baseline understanding of the data.

<b> Check and validate data format </b>

But before we can start describing the data, we need to make sure the data are in their intended formats:
- Check the format of each feature
- Check basic statistics of the dataset

In [21]:
# Check the data format using .info() method
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62642 entries, 0 to 62641
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   timestamp                62642 non-null  object
 1   company                  62642 non-null  object
 2   level                    62642 non-null  object
 3   title                    62642 non-null  object
 4   totalyearlycompensation  62642 non-null  object
 5   location                 62642 non-null  object
 6   yearsofexperience        62642 non-null  object
 7   yearsatcompany           62642 non-null  object
 8   tag                      62642 non-null  object
 9   basesalary               62642 non-null  object
 10  stockgrantvalue          62642 non-null  object
 11  bonus                    62642 non-null  object
 12  gender                   62642 non-null  object
 13  otherdetails             62642 non-null  object
 14  cityid                   62642 non-nul

Notice that each feature is in the format of "object" hence we need to change to appropriate format.

In [22]:
df['timestamp'] = pd.to_datetime(df['timestamp'])

for feature in ['company', 'level', 'cityid', 'dmaid', 'location', 'gender']:
    df[feature] = df[feature].astype('category')

for feature in ['totalyearlycompensation', 'yearsofexperience', 'yearsatcompany', 'basesalary', 'stockgrantvalue', 'bonus']:
    df[feature] = pd.to_numeric(df[feature])
    
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62642 entries, 0 to 62641
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   timestamp                62642 non-null  datetime64[ns]
 1   company                  62642 non-null  category      
 2   level                    62642 non-null  category      
 3   title                    62642 non-null  object        
 4   totalyearlycompensation  62642 non-null  int64         
 5   location                 62642 non-null  category      
 6   yearsofexperience        62642 non-null  int64         
 7   yearsatcompany           62642 non-null  int64         
 8   tag                      62642 non-null  object        
 9   basesalary               62642 non-null  int64         
 10  stockgrantvalue          62642 non-null  int64         
 11  bonus                    62642 non-null  int64         
 12  gender                   62642 n

In [23]:
df.describe(include='all', datetime_is_numeric=True)

timestamp company  level              title  \
count                           62642   62642  62642              62642   
unique                            NaN    1871   3074                 15   
top                               NaN  Amazon     L4  Software Engineer   
freq                              NaN    8054   5008              41231   
mean    2020-07-17 03:30:08.313815040     NaN    NaN                NaN   
min               2017-06-07 11:33:27     NaN    NaN                NaN   
25%        2020-01-11 23:12:55.500000     NaN    NaN                NaN   
50%               2020-09-21 16:31:21     NaN    NaN                NaN   
75%        2021-03-26 15:19:06.500000     NaN    NaN                NaN   
max               2021-08-17 08:28:57     NaN    NaN                NaN   
std                               NaN     NaN    NaN                NaN   

        totalyearlycompensation     location  yearsofexperience  \
count              6.264200e+04        62642       62642.000000   
unique                      NaN         1050                NaN   
top                         NaN  Seattle, WA                NaN   
freq                        NaN         8701                NaN   
mean               3.175327e+03          NaN           7.191054   
min                0.000000e+00          NaN           0.000000   
25%                1.160000e+02          NaN           3.000000   
50%                1.830000e+02          NaN           6.000000   
75%                2.690000e+02          NaN          10.000000   
max                3.386013e+06          NaN          90.000000   
std                3.877877e+04          NaN           6.437148   

        yearsatcompany         tag    basesalary  stockgrantvalue  \
count     62642.000000       62642  6.264200e+04     6.264200e+04   
unique             NaN        3276           NaN              NaN   
top                NaN  Full Stack           NaN              NaN   
freq               NaN       11382           NaN              NaN   
mean          2.700616         NaN  1.798145e+03     1.134865e+03   
min           0.000000         NaN  0.000000e+00     0.000000e+00   
25%           0.000000         NaN  8.800000e+01     0.000000e+00   
50%           1.000000         NaN  1.310000e+02     2.300000e+01   
75%           4.000000         NaN  1.810000e+02     6.400000e+01   
max          90.000000         NaN  2.057772e+06     3.140803e+06   
std           3.524030         NaN  1.927194e+04     2.520611e+04   

                bonus gender otherdetails cityid  dmaid     rowNumber  
count    62642.000000  62642        62642  62642  62642  62642.000000  
unique            NaN      5        40133   1045    150           NaN  
top               NaN   Male               11527    807           NaN  
freq              NaN  35702        22503   8701  20400           NaN  
mean       242.317199    NaN          NaN    NaN    NaN  41694.723732  
min          0.000000    NaN          NaN    NaN    NaN      1.000000  
25%          1.000000    NaN          NaN    NaN    NaN  20069.250000  
50%         13.000000    NaN          NaN    NaN    NaN  42019.000000  
75%         27.000000    NaN          NaN    NaN    NaN  63021.750000  
max     384026.000000    NaN          NaN    NaN    NaN  83875.000000  
std       3709.187697    NaN          NaN    NaN    NaN  24488.865879

Initial glance shows that:
1. Most of the data are from Amazon job roles.
2. Most of the submissions are for level L4, and job title Software Engineer.
3. For totalyearlycompensation, mean and median are quite far away, hence the distribution might be skewed - similarly for basesalary, stockgrantvalue and bonus.
4. Most of the submissions are for job roles located in Seattle, WA.
5. Mean years of experience is around 7 years, with median of 6 years.
6. Mean years at company is 2.7 years, while median is around 1 year - signifying that job hopping is very common for tech roles, where people typically stayed in a company for 1 year.
7. Most of the data are male - showing gender disparity in tech sector.
8. The cityid that appears most is 11527 while 807 for dmaid. Surprisingly, Nielsen's DMA ID only ranges from 1 to 210 hence further checking needed for these features.

Although `df.info()` shows we have no missing values, let's do a further check before we proceed to the next stage. 

Normally, we run `.value_counts()` on each feature, but let's be productive and use the `missingno` library. 

Insights/conclusion from missingno

Now that we know how many values are missing, let's understand more about the data using `pandas-profiling`.

Insights/conclusion from pandas-profiling

### Explore data

Data exploration must be aligned with the project objective - to predict the annual salary (including base pay, stock grants and estimated bonuses) for data science and STEM roles using data from [LEVELS.FYI](https://www.levels.fyi/).

Hence we prioritise the following questions:
1. What are the features that will affect annual compensation offered?
2. How does those features change wrt annual compensation offered?
3. What are the irrelevant (and uncorrelated) features that we can drop?

From here, we can extend to these questions:
1. Many people thinks that tech companies pays well above average and cost of living. Is this true?
2. For the same position and job level, does 2021 compensation higher than in 2017? If yes, is it aligned with inflation?
3. Are women underpaid in these companies? If yes, by how much?
4. Does different company offer differently for the same position or there is no disparity?
5. Does years in company affect total compensation? If yes, which part of compensation was affected - base salary, stock value or bonus?
6. For the same position or job level, does location affect total compensation regardless of gender?

Good to know:
1. What are the factors impacting the salary of a data scientist?
2. What is the salary trend for data scientist? and its comparison to living cost?
3. What is the predicted salary for data scientist in next 5 years?

## References

Best Jobs in America 2022 Employees' Choice (2022). Glassdoor. Retrived on May 3, 2022 from
https://www.glassdoor.com/List/Best-Jobs-in-America-LST_KQ0,20.htm

Data Scientist Salary-The Ultimate Guide for 2021 (2022). ProjectPro. Retrived on May 3, 2022 from https://www.projectpro.io/article/data-scientist-salary-the-ultimate-guide-for-2021/218

Davenport, T H. & Patil, DJ. (2012). Data Scientist: The Sexiest Job of the 21st Century. *HARVARD BUSINESS REVIEW*. Retrieved on 3 May, 2022 from https://hbr.org/2012/10/data-scientist-the-sexiest-job-of-the-21st-century

Martin, I., Mariello, A., Battiti, R. & Hernandez, J A. (2018). Salary Prediction in the IT Job Market with Few High-Dimensional Samples: A Spanish Case Study. *International Journal of Computational Intelligence Systems, 11, 1192-1209*. https://libkey.io/libraries/2611/articles/225996272/full-text-file   
    
    
Miller, H., Seckel, Ea., White, CL., et al. (2021). Gender-based salary differences in academic medicine: a retrospective review of data from six public medical centers in the Western USA. http://dx.doi.org/10.1136/bmjopen-2021-059216
    
Seaman, A. (2021). LinkedIn Jobs on the Rise: 15 opportunities that are in demand and hiring now. Retrieved on May 3, 2022 from https://www.linkedin.com/pulse/linkedin-jobs-rise-15-opportunities-demand-hiring-now-andrew-seaman/